src/Butler/Auth/OIDC.hs

```diff
@@ -0,0 +1,206 @@
+    , uri :: Maybe Text
+    }
+    deriving (Generic, Show)
+
+instance FromJSON OIDCState
+
+instance ToJSON OIDCState
+
+-- | Generate a random fixed size string of 42 char base64 encoded
+genRandomB64 :: IO ByteString
+genRandomB64 = B64.encode . hash <$> genRandom
+
+-- | Generate a random fixed size string of 1024 Bytes
+genRandom :: IO ByteString
+genRandom = do
+    g <- Random.newStdGen
+    let (bs, _ng) = genByteString 1024 g
+    pure bs
+
+mkSessionStore :: OIDCEnv -> Maybe O.State -> Maybe Text -> O.SessionStore IO
+mkSessionStore OIDCEnv{sessionStoreStorage} stateM uriM = do
+    let sessionStoreGenerate = do
+            rb <- liftIO genRandomB64
+            let s = OIDCState (decodeUtf8 rb) uriM
+            pure (B64.encode $ BSL.toStrict $ Data.Aeson.encode s)
+        sessionStoreSave = storeSave
+        sessionStoreGet = storeGet
+        sessionStoreDelete = case stateM of
+            Just state' -> modifyMVar_ sessionStoreStorage $ \store -> pure $ HM.delete state' store
+            Nothing -> pure ()
+     in O.SessionStore{..}
+  where
+    storeSave :: O.State -> O.Nonce -> IO ()
+    storeSave state' nonce = modifyMVar_ sessionStoreStorage $ \store -> pure $ HM.insert state' nonce store
+    storeGet :: O.State -> IO (Maybe O.Nonce)
+    storeGet state' = do
+        store <- readMVar sessionStoreStorage
+        let nonce = HM.lookup state' store
+        pure nonce
+
+loginServer :: OS -> Process -> Sessions -> (Html () -> Html ()) -> JWTSettings -> AuthResult SessionID -> OIDCEnv -> Maybe Workspace -> ServerT LoginAPI Servant.Handler
+loginServer os process sessions mkIndexHtml jwtSettings auth oidcenv mWorkspace =
+    indexRoute auth :<|> loginRoute oidcenv :<|> callbackRoute oidcenv :<|> guestCallbackRoute
+  where
+    callbackRoute :: OIDCEnv -> Maybe Text -> Maybe Text -> Maybe Text -> Servant.Handler AuthResp
+    callbackRoute oidcEnv errM codeM stateM = do
+        case (errM, codeM, stateM) of
+            (Just errorMsg, _, _) -> error $ "Error from remote provider: " <> show errorMsg
+            (_, Nothing, _) -> error "No code parameter given"
+            (_, _, Nothing) -> error "No state parameter given"
+            (_, Just oauthCode, Just oauthState) -> do
+                tokens :: O.Tokens Value <-
+                    liftIO $
+                        O.getValidTokens
+                            (mkSessionStore oidcEnv (Just $ encodeUtf8 oauthState) Nothing)
+                            (oidc oidcEnv)
+                            (manager oidcEnv)
+                            (encodeUtf8 oauthState)
+                            (encodeUtf8 oauthCode)
+                setSession (UserName tokens.idToken.sub)
+
+    guestCallbackRoute :: Servant.Handler AuthResp
+    guestCallbackRoute = setSession "guest"
+
+    setSession :: UserName -> Servant.Handler AuthResp
+    setSession username = do
+        session <- liftIO $ runProcessIO os process $ newSession sessions username
```

Comment:
Perhaps we should lookup the session to check if it doesn't already exist?

Reply:
You mean I use lookupSession to check if the session already exists then if it exists I just change the username ?

Reply:
On the first social login, we can re-use the existing session and simply set its username. However, on further social login (e.g. when the cookie expired, or when using a new browser), we need to recover the existing session, similar to what the recovery link does.

We should rebase on top https://github.com/ButlerOS/haskell-butler/tree/session-provider